examples/shouldi/shouldi/cargo_audit.py

```diff
@@ -0,0 +1,41 @@
+import json
+import asyncio
+from typing import Dict, Any
+
+from dffml.df.base import op
+from dffml.df.types import Definition
+
+package_src_dir = Definition(name="package_src_dir", primitive="str")
+cargo_audit_output = Definition(
+    name="golangci_lint_output", primitive="Dict[str, Any]"
+)
+
+
+class Cargo_AuditError(Exception):
+    """
+    Raised when cargo-audit fails
+    """
+
+
+@op(inputs={"pkg": package_src_dir}, outputs={"report": cargo_audit_output})
+async def run_cargo_audit(pkg: str) -> Dict[str, Any]:
+    """
+    CLI usage: dffml service dev run -log debug shouldi.cargo_audit:run_cargo_audit -pkg .
+    """
+    proc = await asyncio.create_subprocess_exec(
+        "cargo",
+        "audit",
+        "--json",
+        cwd=pkg,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+    )
+    stdout, stderr = await proc.communicate()
+    if len(stdout) == 0:
+        raise Cargo_AuditError(stderr)
+
+    cargo_audit_op = stdout.decode()
```

Comment:
@pdxjohnny I look out for how cargo-audit works, it seems it gives json format file using `cargo-audit audit --json` command but now it is showing a different error 
`raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`

Reply:
Hmm, the GitHub Actions error log is:

```
2020-03-26T19:47:06.8933186Z ======================================================================
2020-03-26T19:47:06.8933533Z ERROR: test_run (tests.test_cargo_audit.TestRunCargo_AuditOp)
2020-03-26T19:47:06.8934501Z ----------------------------------------------------------------------
2020-03-26T19:47:06.8934933Z Traceback (most recent call last):
2020-03-26T19:47:06.8936521Z   File "/home/runner/work/dffml/dffml/dffml/util/asynctestcase.py", line 66, in run_it
2020-03-26T19:47:06.8937194Z     result = self.loop.run_until_complete(coro(*args, **kwargs))
2020-03-26T19:47:06.8937702Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/base_events.py", line 583, in run_until_complete
2020-03-26T19:47:06.8938091Z     return future.result()
2020-03-26T19:47:06.8938473Z   File "/home/runner/work/dffml/dffml/dffml/util/net.py", line 209, in wrapper
2020-03-26T19:47:06.8938859Z     return await func(*(list(args) + [directory_path]), **kwds)
2020-03-26T19:47:06.8939234Z   File "/home/runner/work/dffml/dffml/dffml/util/net.py", line 209, in wrapper
2020-03-26T19:47:06.8939590Z     return await func(*(list(args) + [directory_path]), **kwds)
2020-03-26T19:47:06.8939957Z   File "/home/runner/work/dffml/dffml/dffml/util/net.py", line 209, in wrapper
2020-03-26T19:47:06.8940315Z     return await func(*(list(args) + [directory_path]), **kwds)
2020-03-26T19:47:06.8940690Z   File "/home/runner/work/dffml/dffml/examples/shouldi/tests/test_cargo_audit.py", line 37, in test_run
2020-03-26T19:47:06.8943214Z     / "tarpaulin-59f1c4f48765fba27319cf64e8aab7b08f8a0f66"
2020-03-26T19:47:06.8943506Z   File "/home/runner/work/dffml/dffml/examples/shouldi/shouldi/cargo_audit.py", line 31, in run_cargo_audit
2020-03-26T19:47:06.8943667Z     stderr=asyncio.subprocess.PIPE,
2020-03-26T19:47:06.8943990Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/subprocess.py", line 217, in create_subprocess_exec
2020-03-26T19:47:06.8944157Z     stderr=stderr, **kwds)
2020-03-26T19:47:06.8944356Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/base_events.py", line 1540, in subprocess_exec
2020-03-26T19:47:06.8944502Z     bufsize, **kwargs)
2020-03-26T19:47:06.8944696Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/unix_events.py", line 190, in _make_subprocess_transport
2020-03-26T19:47:06.8944842Z     **kwargs)
2020-03-26T19:47:06.8945128Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/base_subprocess.py", line 37, in __init__
2020-03-26T19:47:06.8945279Z     stderr=stderr, bufsize=bufsize, **kwargs)
2020-03-26T19:47:06.8945457Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/unix_events.py", line 775, in _start
2020-03-26T19:47:06.8945615Z     universal_newlines=False, bufsize=bufsize, **kwargs)
2020-03-26T19:47:06.8945797Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/subprocess.py", line 800, in __init__
2020-03-26T19:47:06.8946366Z     restore_signals, start_new_session)
2020-03-26T19:47:06.8946577Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/subprocess.py", line 1551, in _execute_child
2020-03-26T19:47:06.8946775Z     raise child_exception_type(errno_num, err_msg, err_filename)
2020-03-26T19:47:06.8947185Z FileNotFoundError: [Errno 2] No such file or directory: 'cargo-audit': 'cargo-audit'
2020-03-26T19:47:06.8947281Z 
2020-03-26T19:47:06.8947579Z ----------------------------------------------------------------------
```

It looks like we're downloading the source code to `cargo-audit`, rather than the binary. Therefore, we'll need to first compile `cargo-audit` before we can use it.

You want to make a new variable which is the path to the compiled binary. It will be located at `target/release/cargo-audit` under the `cargo-audit-0.11.2` directory.

If the cargo-audit binary doesn't exist, you'll want to build it by creating a subprocess within the `cargo-audit-0.11.2` directory. The subprocess you want to run is:

```console
$ cargo build --release
```

If that completes successfully, then you're ready to run the test of your operation.

You should this code to compile `cargo-audit` before you run the operation. You'll also want to make sure your adding the `target/release/` directory under `cargo-audit-0.11.2` to the `PATH` rather than just `cargo-audit-0.11.2` (which is what's currently happening).

Reply:
@pdxjohnny should I add the subprocess for running `$ cargo build --release` in the test file or cargo_audit.py file.

Reply:
```python
        cargo_audit_bin_dir = cargo_audit / "cargo-audit-0.11.2" / "target" / "release"
        if not (cargo_audit_bin_dir / "cargo-audit").is_file():
            run_cargo_build(cargo_audit / "cargo-audit-0.11.2")
```

Reply:
```python

        with prepend_to_path(
            rust / "rust-1.42.0-x86_64-unknown-linux-gnu" / "cargo" / "bin",
            cargo_audit_bin_dir,
        ):
```